siphon/simplewebservice/igra2.py

```diff
@@ -0,0 +1,427 @@
+        df_body['date'] = dates_long
+
+        df_body = self._clean_body_df(df_body)
+        df_header = self._clean_header_df(df_header)
+        df_header['date'] = dates
+
+        return df_body, df_header
+
+    def _get_data_raw(self):
+        """Download observations matching the time range.
+
+        Returns a tuple with a string for the body, string for the headers,
+        and a list of dates.
+        """
+        with urlopen(self.ftpsite + self.site_id + self.suffix + '.zip') as url:
+            f = ZipFile(BytesIO(url.read()), 'r').open(self.site_id + self.suffix)
+
+        lines = [line.decode('utf-8') for line in f.readlines()]
+
+        body, header, dates_long, dates = self._select_date_range(lines)
+
+        return body, header, dates_long, dates
+
+    def _select_date_range(self, lines):
+        """Identify lines containing headers within the range begin_date to end_date.
+
+        Parameters
+        -----
+        lines: list
+            list of lines from the IGRA2 data file.
+        """
+        headers = []
+        num_lev = []
+        dates = []
+
+        # Get indices of headers, and make a list of dates and num_lev
+        for idx, line in enumerate(lines):
+            if line[0] == '#':
+                year, month, day, hour = map(int, line[13:26].split())
+
+                # All soundings have YMD, most have hour
+                try:
+                    date = datetime.datetime(year, month, day, hour)
+                except ValueError:
+                    date = datetime.datetime(year, month, day)
+
+                # Check date
+                if self.begin_date <= date <= self.end_date:
+                    headers.append(idx)
+                    num_lev.append(int(line[32:36]))
+                    dates.append(date)
+                if date > self.end_date:
+                    break
+
+        if len(dates) == 0:
+            # Break if no matched dates.
+            # Could improve this later by showing the date range for the station.
+            raise ValueError('No dates match selection.')
+
+        # Compress body of data into a string
+        begin_idx = min(headers)
+        end_idx = max(headers) + num_lev[-1]
+
+        # Make a boolean vector that selects only list indices within the time range
+        selector = np.zeros(len(lines), dtype=bool)
+        selector[begin_idx:end_idx+1] = True
+        selector[headers] = False
+        body = ''.join([line for line in itertools.compress(lines, selector)])
+
+        selector[begin_idx:end_idx + 1] = ~selector[begin_idx:end_idx + 1]
+        header = ''.join([line for line in itertools.compress(lines, selector)])
+
+        # expand date vector to match length of the body dataframe.
+        dates_long = np.repeat(dates, num_lev)
+
+        return body, header, dates_long, dates
+
+    def _get_fwf_params(self):
+        """Produce a dictionary with names, colspecs, and dtype for IGRA2 data.
+        
```

Comment:
Bot is complaining about whitespace on this line--it should be completely blank.